### PR TITLE
Avoid removing the current working directory, even if it becomes empty

### PR DESCRIPTION
**Pull request by newren**
Traditionally, if folks run git commands such as `checkout` or `rebase` from a subdirectory, that git command could remove their current working directory and result in subsequent git and non-git commands either getting confused or printing messages that confuse the user (e.g. "fatal: Unable to read current working directory: No such file or directory").  Many commands either silently avoid removing directories that are not empty (i.e. those that have untracked or modified files in them)[1], or show an error and abort, depending on which is more appropriate for the command in question.  With this series, we augment the reasons to avoid removing directories to include not just has-untracked-or-modified-files, but also to avoid removing the original_cwd as well.

Peff and Junio provided some good pros/cons, if it helps:
  - Pros: Peff (original suggester of the idea)[2], and Junio[3]
  - Cons: Peff [2, again -- see the "P.S."], and Junio[4]

[1] well, with a few exceptions; see https://lore.kernel.org/git/pull.1036.v3.git.1632760428.gitgitgadget@gmail.com/
[2] https://lore.kernel.org/git/YS8eEtwQvF7TaLCb@coredump.intra.peff.net/
[3] https://lore.kernel.org/git/xmqqo86elyht.fsf@gitster.g/
[4] https://lore.kernel.org/git/xmqqo8691gr8.fsf@gitster.g/

Changes since v5:
  - It's been about a week without changes, so including Acks that have come in (let me know if I was wrong in my interpretation of these statements as Acks):
    - from Stolee ("This version is good to go", https://lore.kernel.org/git/aa85e35d-143e-93e4-f54b-146b38dd4b88@gmail.com/)
    - from Ævar ("I've got no objections to these changes going in as they stand.", https://lore.kernel.org/git/211207.86ee6opy0f.gmgdl@evledraar.gmail.com/)

Changes since v4:
  - actually fix bashism

Changes since v3:
  - fixed one codepath from v2 so that the series really is only about the working tree
  - used `test-tool getcwd` instead of `pwd -P` as suggested by Ævar for some less common platforms
  - fixed bashism
  - check for clean index/worktree after verifying that expected-to-abort codepaths do abort, to make it clearer that we expect an early abort
  - remove a leftover (and confusing) is_absolute_dir() check in sequencer and stash from an earlier round of the series

Changes since v2:
  - the series is now only about the working tree.  So if the original cwd is outside the worktree (or we're in a bare repo), then the new code is a no-op.
  - fixed ugly early die() possibility (uses strbuf_getcwd() instead of xgetcwd())
  - modified the initial tests to show both expected and desired behavior.  subsequent patches fix the tests.  One new patch added at the end which simplifies the tests to only check for desired behavior.
  - NULLify startup_info->original_cwd when it matches the toplevel worktree; that is already protected and we don't need secondary protection for it.  This simplified some other codepaths so we don't have to check for startup_info->original_cwd == "".
  - clarified some commit messages

Changes since v1:
  - clarified multiple commit messages
  - renamed the_cwd to startup_info->original_cwd to make it clearer that it's our parent process'es cwd that really matters, which we inherited at program startup.  Also pulls it out of the global namespace.
  - Normalize the path for startup_info->original_cwd, and ensure that it's actually the original cwd even if -C is passed to git.
  - small code cleanups suggested by René and Ævar
  - split the final patch (which got the most comments) into two -- one for each function being modified.  Also, add a bunch more history to the first of the two resulting commit messages
  - no longer has a content conflict with so/stash-staged
  - add another value for the flags parameter that remove_dir_recursively() takes so that it can opt into either the old or the new behavior.  Use that for the one special corner case I could find where it matters, and add a few tests around it to highlight the utility of the flag.

cc: Jeff King <peff@peff.net>
cc: René Scharfe <l.s.r@web.de>
cc: Ævar Arnfjörð Bjarmason <avarab@gmail.com>
cc: Elijah Newren <newren@gmail.com>
cc: Glen Choo <chooglen@google.com>
cc: Philip Oakley <philipoakley@iee.email>
cc: Derrick Stolee <stolee@gmail.com>
cc: Eric Sunshine <sunshine@sunshineco.com>
cc: Phillip Wood <phillip.wood123@gmail.com>

### CODE REVIEW

**Comment by Yuskabss** [comment]
Yandex.com

Android için Outlook<https://aka.ms/AAb9ysg>’u edinin
________________________________

**Comment by Yuskabss** [comment]
Yandex

Android için Outlook<https://aka.ms/AAb9ysg>’u edinin
________________________________

**Comment by Yuskabss** [comment]
Yandex.ru

29 Kas 2021 Pzt 23:00 tarihinde gitgitgadget-git[bot] <
***@***.***> şunu yazdı:

> ***@***.***[bot]* commented on this pull request.
> ------------------------------
>
> In cache.h <https://github.com/git/git/pull/1140#discussion_r758801644>:
>
> > @@ -1834,8 +1834,10 @@ void overlay_tree_on_index(struct index_state *istate,
>
>  struct startup_info {
>
>
> On the Git mailing list
> ***@***.***>,
> Elijah Newren wrote (reply to this
> <https://github.com/gitgitgadget/gitgitgadget/wiki/ReplyToThis>):
>
> On Mon, Nov 29, 2021 at 12:41 PM Ævar Arnfjörð Bjarmason
>

**Comment by Eradnavna** [comment]
I  have no access to this account anymore. Please stop email me. On Tue, 30 Nov 2021 at 00:44, gitgitgadget-git[bot] ***@***.***> wrote:
User Eric Sunshine ***@***.***> has been added to the cc: list.

?You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.
